source/_posts/2024-08-19-state-of-generics-and-collections.md

```diff
@@ -0,0 +1,437 @@
+- Generic type information is not enforced at runtime (so if you don't run a static analysis tool in advance, they accomplish nothing).
+
+## Erased Generic Type Declarations
+
+Seeing the difficulties with implementing reified generics in PHP core, it has been proposed to implement them only syntactically, leaving the task of type checking to static analysers.
+
+In this alternative, the PHP syntax would be changed so that type declarations, class declarations, and function declarations accept generic syntax, but the PHP engine would not check them.
+
+We can call these “Erased” type declarations because the engine will simply ignore them at runtime. This alternative could be implemented in various ways:
+
+- As part of php-src
+- In an extension
+- In an autoloader
+- etc
+
+Here is what the same Dict class as above would look like:
+
+	class Dict<Key,Value>
+	{
+		public function __construct(private array<Key,Value> $entries) {}
+
+		public function set(Key $key, Value $value): self
+		{
+			$this->entries[$key] = $value;
+			return $this;
+		}
+	}
+
+	function f(Dict<Key,Value> $dict) {}
+
+	$dict = new Dict([1 => 'foo']);
+	$dict->set("foo", "bar"); // Static analyser error
+	$dict->set(1, "bar");     // Ok
+	f($dict);                 // Static analyser error
+
+This addresses the docblock verbosity issue of the Static Analysis alternative, but this introduces an inconsistency: Type declarations cause coercion, but erased generic type declarations will not.
+
+Consider the following example:
+
+	class StringList
+	{
+		public function add(string $value)
+		{
+			$this->values[] = $value;
+		}
+	}
+
+	class List<T>
+	{
+		public function add(T $value)
+		{
+			$this->values[] = $value;
+		}
+	}
+
+	$list = new StringList();
+	$list->add(123); // coerced to string
+
+	$list = new List<string>();
+	$list->add(123); // NOT coerced to string
+
+In this scenario, the first call to `add()` will coerce the argument to string, but not the second one.
+
+In languages like Java, which has erased generics on top of a traditional type system, the compiler does type checking, so inconsistencies like the above do not exist. However, in PHP these are unavoidable.
+
+Another drawback of erased generics is they are not visible at runtime. This would prevent pattern matching from seeing generic type arguments, for example.
+
+## Fully Erased Type Declarations
+
+One way to address the inconsistency issue of Erased Generics is to change all type declarations to be erased. This could be opt-in with a `declare()` statement:
+
+	declare(types=erased);
+
+In this alternative, the engine would stop checking types at runtime. In the previous example, both calls to `add()` would have the same behaviour: the value is not coerced. It is then up to the user to check types with an analyser.
+
+This is not uncommon in mainstream interpreted languages, as all of Javascript (via TypeScript), Python, and Ruby have fully erased type declarations.
+
```

Comment:
```suggestion

Letting users opt-in for fully erased typing and generics on a per-file basis would be a trivial change in core and would make working with PHPStan/Psalm generics less verbose. Some additional benefits of this solution include:
* Short-term performance improvements due to the lack of type checking at runtime in opted-in code.
* The potential for extending the type system with advanced types like non-empty-string, list, int<range>, class-string, conditional types, and more.
```

Reply:
I don't think it should talk about "trivial change", as nothing is ever as trivial as it seems, and it makes it sounds this is really easy™ to implement.

Instead of `a per-file basis would be a trivial change in core and would`, perhaps `a per-file basis would be another option to look at, and would`

Reply:
Agreed

Reply:
I've committed and pushed that now.